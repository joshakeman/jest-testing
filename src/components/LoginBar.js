import React, {Component} from 'react'
import axios from 'axios'
import { Button, Navbar } from 'react-bootstrap';
import './../App.css';

class LoginBar extends Component {
    constructor(props) {
        super(props);
        this.state = {username: '', password: '', nickname: '', loggedIn: false}
    }

    handleChange = () => {
        return (event) => {
            let key = event.target.getAttribute('name');
            let newState = {...this.state};
            newState[key] = event.target.value;
            this.setState(newState);
        }
    };

    handleLogin = (event) => {
        event.preventDefault();
        return axios.post(this.props.url + '/active_user', {username: this.state.username, password: this.state.password})
            .then((data) => {
                this.props.onLogin(data.data.user_auth)}
                )
            .then(this.setState({...this.state, loggedIn: true}))
            .catch(e => console.log(e));

    };

    handleRegister = (event) => {
        event.preventDefault();
        return axios.post(this.props.url + '/user', {username: this.state.username, password: this.state.password, handle: this.state.nickname})
            .then((data) => this.handleLogin(event))
            .catch(e => console.log(e));

    };

    render = () => {
        return (
            <Navbar className="App-header">
                <Navbar.Header>
                    <img className="App-logo" width={80} height={40} src="/images/logo.png" alt="logo"/>
                    <Navbar.Toggle />
                </Navbar.Header>
                {this.state.loggedIn ? "" :
                <Navbar.Collapse>
                    <Navbar.Form pullLeft>
                        <label>
                            Username:
                            <input name="username" type="text" value={this.state.username}
                                   onChange={this.handleChange()}/>
                        </label>
                        <label>
                            Password:
                            <input name="password" type="password" value={this.state.password}
                                   onChange={this.handleChange()}/>
                        </label>
                        <label>
                            Nickname:
                            <input name="nickname" type="text" value={this.state.nickname}
                                   onChange={this.handleChange()}/>
                        </label>
                        <Button className="App-Button" bsSize="small" type="submit" onClick={this.handleLogin}>Login</Button>
                        <Button className="App-Button" bsSize="small" type="submit" onClick={this.handleRegister}>Register</Button>
                    </Navbar.Form>
                </Navbar.Collapse>}
            </Navbar>
        )
    }
}

export default LoginBar
