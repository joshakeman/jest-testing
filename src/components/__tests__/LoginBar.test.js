/**
 * @jest-environment jsdom
 */

import React from 'react'
import ReactDOM from 'react-dom'
import axios from 'axios'

jest.mock('axios');

import LoginBar from '../LoginBar'


describe('LoginBar tests', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<LoginBar/>, div);
        ReactDOM.unmountComponentAtNode(div)
    });

    it('shows 2 buttons when logged out', () => {
        const loginBar = shallow(<LoginBar/>);
        expect(loginBar.find('.App-logo').length).toBe(1);
        expect(loginBar.find('.App-Button').length).toBe(2)
    });

    it('matches snapshot when logged out', () => {
        const loginBar = shallow(<LoginBar/>);
        expect(loginBar).toMatchSnapshot();
    });

    it('shows only logo when logged in', () => {
        const loginBar = shallow(<LoginBar/>);
        loginBar.setState({...loginBar.state, loggedIn: true});
        expect(loginBar.find('.App-logo').length).toBe(1);
        expect(loginBar.find('.App-Button').length).toBe(0)
    });

    it('matches snapshot when logged in', () => {
        const loginBar = shallow(<LoginBar/>);
        loginBar.setState({...loginBar.state, loggedIn: true});
        expect(loginBar).toMatchSnapshot();
    });



    it('Calls onLogin with proper credentials when login is clicked', async () => {
        // Prepare the test
        const onLoginMock = jest.fn();
        const ok = {data: {user_auth: "OK"}};
        let post = axios.post;
        post.mockImplementation((url, payload) => Promise.resolve(ok));
        let loginBar = shallow(<LoginBar url='' onLogin={onLoginMock}/>).instance();
        loginBar.setState({...loginBar.state, username: 'Bob', password: 'secret'});

        // Make sure all our assertions are actually called and the test doesn't bail out early
        expect.assertions(6);

        // Invoke the code under test - the handleLogin() function - and wait for it to finish
        await loginBar.handleLogin({preventDefault: jest.fn()});

        // Verify axios was called with the proper parameters
        expect(post.mock.calls.length).toBe(1);
        expect(post).toHaveBeenCalledTimes(1);
        const expectedArgs = ['/active_user', {username: 'Bob', password: 'secret'}];
        expect(post).toHaveBeenCalledWith(...expectedArgs);

        // Verify handleLogin() properly called the onLogin() callabck
        expect(onLoginMock.mock.calls.length).toBe(1);
        expect(onLoginMock).toHaveBeenCalledTimes(1);
        expect(onLoginMock).toHaveBeenCalledWith('OK');
    })
});