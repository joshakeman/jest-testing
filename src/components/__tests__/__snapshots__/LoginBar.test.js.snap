// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginBar tests matches snapshot when logged in 1`] = `
<Uncontrolled(Navbar)
  bsStyle="default"
  className="App-header"
>
  <NavbarHeader>
    <img
      alt="logo"
      className="App-logo"
      height={40}
      src="/images/logo.png"
      width={80}
    />
    <NavbarToggle />
  </NavbarHeader>
</Uncontrolled(Navbar)>
`;

exports[`LoginBar tests matches snapshot when logged out 1`] = `
<Uncontrolled(Navbar)
  bsStyle="default"
  className="App-header"
>
  <NavbarHeader>
    <img
      alt="logo"
      className="App-logo"
      height={40}
      src="/images/logo.png"
      width={80}
    />
    <NavbarToggle />
  </NavbarHeader>
  <NavbarCollapse>
    <NavbarForm
      componentClass="div"
      pullLeft={true}
      pullRight={false}
    >
      <label>
        Username:
        <input
          name="username"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <label>
        Password:
        <input
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </label>
      <label>
        Nickname:
        <input
          name="nickname"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="small"
        bsStyle="default"
        className="App-Button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Login
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="small"
        bsStyle="default"
        className="App-Button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Register
      </Button>
    </NavbarForm>
  </NavbarCollapse>
</Uncontrolled(Navbar)>
`;
